## Learning Statistics with Python: Central Tendency and Data Dispersion

In [ ]:
#Install the libraries into your environment
#pip install numpy
#pip install scipy

In [1]:
# Suppose we have a sample data
data = [2, 4, 6, 8, 10, 4, 7, 3, 5]

## 1. Measures of Central Tendency
### Using Pure Python

In [2]:
# Mean
mean = sum(data) / len(data)

In [3]:
# Median
data_sorted = sorted(data)
n = len(data)
if n % 2 == 0:
    median = (data_sorted[n//2 - 1] + data_sorted[n//2]) / 2
else:
    median = data_sorted[n//2]

In [4]:
# Mode
from collections import Counter
data_count = Counter(data)
mode = [k for k, v in data_count.items() if v == max(data_count.values())]

print("Mean:", mean)
print("Median:", median)
print("Mode:", mode)

Mean: 5.444444444444445
Median: 5
Mode: [4]


### Using NumPy and SciPy

In [11]:
import numpy as np
from scipy import stats

mean_np = np.mean(data)
median_np = np.median(data)
mode_np = stats.mode(data).mode

print("Mean (NumPy):", mean_np)
print("Median (NumPy):", median_np)
print("Mode (SciPy):", mode_np)

Mean (NumPy): 5.444444444444445
Median (NumPy): 5.0
Mode (SciPy): 4


## 2. Measures of Dispersion
### Using Pure Python

In [12]:
# Range= Maximum value - Minimum value
range_data = max(data) - min(data)

In [13]:
# Variance
variance_population = sum((x - mean)**2 for x in data) / len(data)
variance_sample = sum((x - mean)**2 for x in data) / (len(data)-1)


In [14]:
# Standard Deviation
std_population = variance_population**0.5
std_sample = variance_sample**0.5

In [15]:
print("Range:", range_data)
print("Population Variance:", variance_population)
print("Sample Variance:", variance_sample)
print("Population SD:", std_population)
print("Sample SD:", std_sample)

Range: 8
Population Variance: 5.802469135802469
Sample Variance: 6.527777777777778
Population SD: 2.4088314876309775
Sample SD: 2.5549516194593154


### Using NumPy

In [16]:
variance_np_population = np.var(data)
variance_np_sample = np.var(data, ddof=1)
std_np_population = np.std(data)
std_np_sample = np.std(data, ddof=1)

print("Population Variance (NumPy):", variance_np_population)
print("Sample Variance (NumPy):", variance_np_sample)
print("Population SD (NumPy):", std_np_population)
print("Sample SD (NumPy):", std_np_sample)

Population Variance (NumPy): 5.80246913580247
Sample Variance (NumPy): 6.527777777777779
Population SD (NumPy): 2.408831487630978
Sample SD (NumPy): 2.5549516194593154


### Interquartile Range (IQR) Using NumPy

In [17]:
Q1 = np.percentile(data, 25)
Q3 = np.percentile(data, 75)
IQR = Q3 - Q1

print("Q1:", Q1)
print("Q3:", Q3)
print("Interquartile Range (IQR):", IQR)

Q1: 4.0
Q3: 7.0
Interquartile Range (IQR): 3.0


After gaining a basic understanding of these concepts, let's apply the same ideas to real-world data using pandas. Seaborn is a Python library for data visualization that includes several built-in example datasets, such as Titanic, Iris, Tips, and Penguins. In this example, we will focus on the Iris dataset, which contains measurements of various iris flower species.

In [19]:
import seaborn as sns

# Load the Iris dataset
iris = sns.load_dataset('iris')
print(iris.head())


   sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa


In [22]:
# Select numeric columns only
numeric_cols = ['sepal_length', 'sepal_width', 'petal_length', 'petal_width']
iris_numeric = iris[numeric_cols]

In [24]:
# Central Tendency; mean, median, mode
mean_df = iris_numeric.mean()
median_df = iris_numeric.median()
mode_df = iris_numeric.mode().iloc[0]  # take the first row in case of multiple modes
print("Mean (DataFrame):\n", mean_df)
print("Median (DataFrame):\n", median_df)
print("Mode (DataFrame):\n", mode_df)


Mean (DataFrame):
 sepal_length    5.843333
sepal_width     3.057333
petal_length    3.758000
petal_width     1.199333
dtype: float64
Median (DataFrame):
 sepal_length    5.80
sepal_width     3.00
petal_length    4.35
petal_width     1.30
dtype: float64
Mode (DataFrame):
 sepal_length    5.0
sepal_width     3.0
petal_length    1.4
petal_width     0.2
Name: 0, dtype: float64


In [26]:
# Dispersion
range_df = iris_numeric.max() - iris_numeric.min()
variance_df = iris_numeric.var()      # sample variance by default (ddof=1)
std_df = iris_numeric.std()           # sample standard deviation
iqr_df = iris_numeric.quantile(0.75) - iris_numeric.quantile(0.25)
print("Range (DataFrame):\n", range_df)
print("Variance (DataFrame):\n", variance_df)
print("Standard Deviation (DataFrame):\n", std_df)
print("Interquartile Range (DataFrame):\n", iqr_df)

Range (DataFrame):
 sepal_length    3.6
sepal_width     2.4
petal_length    5.9
petal_width     2.4
dtype: float64
Variance (DataFrame):
 sepal_length    0.685694
sepal_width     0.189979
petal_length    3.116278
petal_width     0.581006
dtype: float64
Standard Deviation (DataFrame):
 sepal_length    0.828066
sepal_width     0.435866
petal_length    1.765298
petal_width     0.762238
dtype: float64
Interquartile Range (DataFrame):
 sepal_length    1.3
sepal_width     0.5
petal_length    3.5
petal_width     1.5
dtype: float64


In [28]:
# Combine all statistics into a single DataFrame
import pandas as pd
iris_stats_df = pd.DataFrame({
    'Mean': mean_df,
    'Median': median_df,
    'Mode': mode_df,
    'Range': range_df,
    'Variance': variance_df,
    'Standard Deviation': std_df,
    'IQR': iqr_df
})

# Display the result
print(iris_stats_df)

                  Mean  Median  Mode  Range  Variance  Standard Deviation  IQR
sepal_length  5.843333    5.80   5.0    3.6  0.685694            0.828066  1.3
sepal_width   3.057333    3.00   3.0    2.4  0.189979            0.435866  0.5
petal_length  3.758000    4.35   1.4    5.9  3.116278            1.765298  3.5
petal_width   1.199333    1.30   0.2    2.4  0.581006            0.762238  1.5


### How to interpret the results?
- Range (max - min): Shows the spread from smallest to largest values. Petal length has the largest range (5.9), while sepal width is least variable (2.4).
- Variance: Measures how spread out values are around the mean. Petal length has the highest variance (3.11), sepal width the lowest (0.19).
- Standard Deviation (SD): Petal_length has SD ≈ 1.76, meaning on average, values deviate ~1.76 units from the mean. Sepal_width SD ≈ 0.44, showing it’s very consistent.
- Interquartile Range (IQR): Petal_length IQR ≈ 3.55 → middle 50% of petal lengths spans 3.55 units.
Sepal_width IQR ≈ 0.5 → middle 50% of sepal widths is tightly packed.